### PR TITLE
EPMRPP-33316 React. Settings. BTS settings

### PR DESCRIPTION
**Pull request by AmsterGet**

### CODE REVIEW

File: app/src/pages/inside/profilePage/configExamplesBlock/configExamplesBlock.jsx [hunk 1]
```diff
@@ -41,6 +41,17 @@ export class ConfigExamplesBlock extends Component {
   static defaultProps = {
     token: '',
   };
+
+  state = {
+    activeTab: 0,
+  };
+
+  tabChangeHandle = (activeTab) => {
```
Comment (DiscoElevator) [inline app/src/pages/inside/profilePage/configExamplesBlock/configExamplesBlock.jsx]: tabChangeHandler

File: app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/jiraInstances/jiraInstances.jsx [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+  removeTabHandle = () => {
+    const { projects, prevActive } = this.state;
+    this.setState({
+      projects: [...projects.slice(0, projects.length - 1)],
+      activeInstance: prevActive,
+    });
+  };
+
+  newProjectInstance = (removable) => ({
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/jiraInstances/jiraInstances.jsx]: getNewProjectInstance

File: app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/modals/deleteInstanceModal/deleteInstanceModal.jsx [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    externalSystems: PropTypes.array.isRequired,
+    data: PropTypes.object,
+    showNotification: PropTypes.func.isRequired,
+    updateExternalSystemAction: PropTypes.func.isRequired,
+  };
+  static defaultProps = {
+    data: {},
+  };
+
+  onClickDelete = (closeModal) => {
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/modals/deleteInstanceModal/deleteInstanceModal.jsx]: deleteInstance

File: app/src/controllers/project/constants.js [hunk 4]
```diff
@@ -7,3 +7,4 @@ export const FETCH_PROJECT_PREFERENCES_SUCCESS = 'fetchProjectPreferencesSuccess
 export const TOGGLE_DISPLAY_FILTER_ON_LAUNCHES = 'toggleDisplayFilterOnLaunches';
 export const UPDATE_AUTO_ANALYSIS_CONFIGURATION = 'updateAutoAnalysisConfigurationSuccess';
 export const UPDATE_EMAIL_CONFIG_SUCCESS = 'updateEmailConfigSuccess';
+export const UPDATE_EXTERNAL_SYSTEM = 'updateExternalSystemSuccess';
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply app/src/controllers/project/constants.js]: Fixed

File: app/src/pages/inside/common/dynamicFieldsBlock/dynamicFieldsBlock.jsx [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+  };
+
+  static defaultProps = {
+    fields: [],
+    onFieldCheck: null,
+    withFormValidation: false,
+  };
+
+  getInputOptions = (values) =>
+    ((values[0] && values) || []).map((item) => ({ value: item.valueId, label: item.valueName }));
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply app/src/pages/inside/common/dynamicFieldsBlock/dynamicFieldsBlock.jsx]: Fixed

File: app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/jiraInstances/jiraInstances.jsx [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        instancesQuantity: projects.length,
+        activeInstance,
+      };
+    }
+    return null;
+  }
+
+  shouldComponentUpdate(nextProps, nextState) {
+    !nextState.projects.length &&
+      this.setState({
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/jiraInstances/jiraInstances.jsx]: Remove it method. Moved this logic into `getDerivedStateFromProps` for update component state when it hasn't any projects

File: app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/jiraInstances/jiraInstances.jsx [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+  };
+
+  static defaultProps = {
+    instances: [],
+  };
+
+  static createTabsConfig = (instances) =>
+    instances.map(({ project, ...rest }) => ({
+      name: project,
+      content: <JiraInstanceForm instanceData={{ ...rest, project, password: '11111111' }} />,
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/jiraInstances/jiraInstances.jsx]: In backbone version in this password field `11111111` value renders as a mock value.
Can I write it as a constant like `MOCK_PASSWORD`?

[the remark this answers is not in the document]
Comment (DiscoElevator) [reply app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/jiraInstances/jiraInstances.jsx]: Yes

File: app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/jiraInstances/jiraInstances.jsx (hunk 2 above)
Comment (AmsterGet) [reply app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/jiraInstances/jiraInstances.jsx]: Fixed

File: app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/modals/deleteInstanceModal/deleteInstanceModal.jsx (hunk 3 above)
Comment (AmsterGet) [reply app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/modals/deleteInstanceModal/deleteInstanceModal.jsx]: Fixed

File: app/src/pages/inside/profilePage/configExamplesBlock/configExamplesBlock.jsx (hunk 1 above)
Comment (AmsterGet) [reply app/src/pages/inside/profilePage/configExamplesBlock/configExamplesBlock.jsx]: Fixed

File: app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/jiraInstances/jiraInstanceForm/jiraInstanceForm.jsx [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+
+  setDisabled = (value) => {
+    this.setState({
+      disabled: value,
+    });
+  };
+
+  parseInputValueHandler = (value) => {
+    if (value !== this.state.instanceData.project || value !== this.state.instanceData.url) {
+      this.props.change('username', '');
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/jiraInstances/jiraInstanceForm/jiraInstanceForm.jsx]: Fixed anywhere

File: app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/jiraInstances/jiraInstanceForm/jiraInstanceForm.jsx [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+      };
+    }
+    return null;
+  }
+
+  onCancel = () => {
+    this.setState({
+      disabled: true,
+    });
+    this.props.initialize(JiraInstanceForm.mapInstanceDataToFormData(this.props.instanceData));
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/jiraInstances/jiraInstanceForm/jiraInstanceForm.jsx]: Done

File: app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/defaultPropertiesForIssueForm/defaultPropertiesForIssueForm.jsx [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+      instanceData: props.instanceData,
+      fields: normalizedFields,
+      issueType: '',
+      fieldsPrepared: false,
+      editMode: false,
+    };
+  }
+
+  static getDerivedStateFromProps(props, state) {
+    if (JSON.stringify(state.instanceData) !== JSON.stringify(props.instanceData)) {
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/defaultPropertiesForIssueForm/defaultPropertiesForIssueForm.jsx]: Done

File: app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/jiraInstances/jiraInstanceForm/jiraInstanceForm.jsx [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+      disabled: !props.newItem,
+    };
+  }
+
+  componentDidMount() {
+    this.props.initialize(JiraInstanceForm.mapInstanceDataToFormData(this.props.instanceData));
+  }
+
+  static getDerivedStateFromProps(props, state) {
+    if (JSON.stringify(state.instanceData) !== JSON.stringify(props.instanceData)) {
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/jiraInstances/jiraInstanceForm/jiraInstanceForm.jsx]: Done

File: app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/jiraInstances/jiraInstanceForm/jiraInstanceForm.jsx [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+      props.initialize(JiraInstanceForm.mapInstanceDataToFormData(props.instanceData));
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/jiraInstances/jiraInstanceForm/jiraInstanceForm.jsx]: Moved it into `componentDidUpdate`

File: app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/jiraInstances/jiraInstanceForm/jiraInstanceForm.jsx [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+      return {
+        instanceData: props.instanceData,
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/jiraInstances/jiraInstanceForm/jiraInstanceForm.jsx]: I do it to compare previous `instanceData` with new.
Thinking about better solution...

File: app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/defaultPropertiesForIssueForm/defaultPropertiesForIssueForm.jsx [hunk 14]
```diff
@@ -0,0 +1,377 @@
...
+
+  checkIfTheSystemHasFields = () =>
+    this.props.instanceData.fields && !!this.props.instanceData.fields.length;
+
+  prepareFieldsToRender = () => {
+    const {
+      instanceData: { fields },
+      initialize,
+    } = this.props;
+    const updatedFields = this.state.fields;
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply app/src/pages/inside/settingsPage/settingTabs/bugTrackingSystemTab/instancesBlock/defaultPropertiesForIssueForm/defaultPropertiesForIssueForm.jsx]: Fixed

**Comment by AmsterGet** [comment]
Aslo rewrote `FormField` component to use it in the `DynamicFieldsBlock`

**Comment by AmsterGet** [comment]
Blocked by backend

**Pull request closed without merging**